node-neigh: Locking, logging, misc improvements

brb opened the pull request:
See commit msgs.

/cc @jaffcheng 

Fix #15676

brb committed:
node-neigh: Introduce dedicated lock

This commit introduces a dedicated lock for managing node neighbor
entries.

Previously, the neighbor subsystem was sharing the same lock with the
node manager. This resulted in the lock starvation when the periodic
neighbor refresh was taking longer than expected and it was queuing up
the refresh requests.

To avoid the starvation which can slow down node updates, we introduce
the dedicated lock. Also, we make sure that insertNeighbor() won't block
NodeUpdate() in the case of the scenario above by running it in a
separate goroutine (same applies to deleteNeighbor() and NodeDelete()).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
node-neigh: Remove unnecessary looping over recv

There is no default case in the select, so the select will block until
one of its cases is fulfilled.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
node-neigh: Misc refactoring

- Lock insertNeighbor() before accessing shared structures.
- Defer neighbor BPF map retirement so that it's moved from the contend
  path.
- Make getSrcAndNextHopIPv4() a function to indicate that it's not using
  any shared structures.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
node-neigh: Make neighLock'ing more fine-grained

This commit changes how insertNeighbor() method takes the lock. Instead
of holding the lock for the entire execution duration, we release the
lock while arpinging. This allows parallel arpings which should make the
periodic ARP refresher faster.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
node-neigh: Do not return early if nextHop is the same

It can happen that the insertNeighbor(refresh=true) gets called before
insertNeighbor(refresh=false), and only the latter can increment the
neigh refcounting.

Remove the optimization for the sake of refcounting correctness.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
daemon: Add --arping-refresh-period

Instead of deriving the value from
net.ipv4.neigh.default.base_reachable_time_ms which is 30s by default
use a user specified value for the ARP periodic refresher.

Running ARP refresher every 15-45s is too often, as in most clusters
node hw addrs don't change at all.

Users who have highly dynamic environments will be able to schedule the
refresher in higher frequency by using the flag.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro commented on pkg/datapath/linux/node.go:
```diff
@@ -902,8 +899,8 @@ func (n *linuxNodeHandler) nodeUpdate(oldNode, newNode *nodeTypes.Node, firstAdd
-	if n.enableNeighDiscovery {
-		n.insertNeighbor(context.Background(), newNode, false)
+	if n.enableNeighDiscovery && !newNode.IsLocal() {
+		go n.insertNeighbor(context.Background(), newNode, false)
```
very tiny nit: I think this `go` statement here makes two assumptions. Both of which seem valid, but maybe we should leave a comment about it here.

1. `newNode` is now accessed concurrently by multiple go routines. All access seem to be reads (I haven't found any writes), so it seems safe.
2. It assumes it is safe to invoke `insertNeighbor` for the same node concurrently. Reading the rest of the commits, this also seems safe, but I'm less confident here.

brb replied:
Yep, makes sense, I will add it.

gandro commented on pkg/datapath/linux/node.go:
```diff
@@ -680,15 +677,18 @@ func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeType
 		logfields.Interface: n.neighDiscoveryLink.Attrs().Name,
```
I believe you have a race here. `n.neighDiscoveryLink` may be updated concurrently:
https://github.com/cilium/cilium/blob/1309d2aa936f50b73ceb7e47458260ae9d7e8b15/pkg/datapath/linux/node.go#L1364

Setting a pointer in one go routine and reading it in another without synchronization is a data race I believe. Before this PR, this code here safe because `insertNeighbor` held the `n.mutex` lock. 

You also have more unprotected accesses to `n.neighDiscoveryLink` below:
https://github.com/cilium/cilium/blob/1309d2aa936f50b73ceb7e47458260ae9d7e8b15/pkg/datapath/linux/node.go#L743

brb replied:
Good catch! Probably I should use `sync/atomic.{Load,Swap}Pointer()` for `n.neighDiscoveryLink`, and return early in the `insertNeighbor()` if its nil. UPDATED: Link is an interface, so I'm going to introduce a bool to indicate whether the link has been set.

brb replied:
Actually it is enough just to check whether `neighDiscoveryLink` is set in `insertNeighbor()`.

gandro commented on pkg/node/manager/manager.go:
```diff
@@ -571,14 +547,12 @@ func (m *Manager) StartNeighborRefresh(nh datapath.NodeHandler) {
 						continue
 					}
 					go func(c context.Context, e nodeTypes.Node) {
-						n := randGen.Int63n(int64(interval / 2))
-						time.Sleep(interval/2 + time.Duration(n))
```
I agree with the removal of the `net.ipv4.neigh.default.base_reachable_time_ms` based-interval, but this sleep statement here seems to introduce some randomness such that we don't ping all nodes at once, do we want to keep it?

brb replied:
Yeah, I was thinking to add this in a separate PR to avoid overcomplicating this one so that reviewers could focus on concurrency aspects of this PR instead. But anyway, I will add it.

gandro reviewed (commented):
I'm not a high-confidence reviewer in this part of the code, so I did a review solely based on the patch but I have a few questions. I will do another in-depth review later.

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -61,6 +61,7 @@ type linuxNodeHandler struct {
 	neighNextHopByNode   map[nodeTypes.Identity]string // val = string(net.IP)
 	neighNextHopRefCount counter.StringCounter
 	neighByNextHop       map[string]*netlink.Neigh // key = string(net.IP)
+	neighLock            lock.Mutex
```
Can you leave a comment describing what is this mutex protecting?

aanm replied:
unresolved.

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -682,30 +677,31 @@ func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeType
+		if locked {
+			n.neighLock.Unlock()
+		}
+	}()
```
you can remove this

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -728,16 +724,25 @@ func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeType
 		metrics.ArpingRequestsTotal.WithLabelValues(success).Inc()
+	}
 
+	n.neighLock.Lock()
```
and you can add the defer here:
```
	defer n.neighLock.Unlock()
```

brb committed:
node-neigh: Log arping failures only in debug mode

Most of arping failures which we observed are transient, so these false
positives might mislead users.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

gandro commented on pkg/datapath/linux/node.go:
```diff
@@ -729,16 +743,25 @@ func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeType
+	var hwAddr net.HardwareAddr
 	if nextHopIsNew || refresh {
-		hwAddr, err := arp.PingOverLink(n.neighDiscoveryLink, srcIPv4, nextHopIPv4)
+		hwAddr, err = arp.PingOverLink(n.neighDiscoveryLink, srcIPv4, nextHopIPv4)
```
I don't believe this is safe yet, i.e. the race detect can still complain about this. I believe you need to copy `n.neighDiscoveryLink` into a local variable while holding the `n.neighLock.Lock()`

The reason being that `n.neighDiscoveryLink` can be set more than once, i.e. when `cilium config` is called, then `NodeConfigurationChanged` is invoked again, setting `n.neighDiscoveryLink` again even if it wasn't nil before:
https://github.com/cilium/cilium/blob/e5fbe232dcbdb0ae63ba8a93a289a115e7f29afd/daemon/cmd/config.go#L101

gandro reviewed (approved):
:rocket:

jaffcheng commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -18,6 +18,7 @@ cilium-agent [flags]
       --allow-localhost string                               Policy when to allow local stack to reach local endpoints { auto | always | policy } (default "auto")
       --annotate-k8s-node                                    Annotate Kubernetes node (default true)
       --api-rate-limit map                                   API rate limiting configuration (example: --rate-limit endpoint-create=rate-limit:10/m,rate-burst:2) (default map[])
+      --arping-refresh-period duration                       Period for ARP entries to remote nodes refresher (default 5m0s)
```
Does it make sense to make it possible to disable refresh completely if hw addrs don't change in most clusters? E.g. give a zero or negative duration?

brb replied:
That's a good idea, will do that!

jaffcheng reviewed (commented):
Thanks for fixing this, just one non-blocking comment, feel free to ignore.

twpayne reviewed (approved):
The CLI parts LGTM.

joamaki commented on daemon/cmd/daemon_main.go:
```diff
@@ -230,6 +230,9 @@ func init() {
 	flags.Bool(option.AnnotateK8sNode, defaults.AnnotateK8sNode, "Annotate Kubernetes node")
 	option.BindEnv(option.AnnotateK8sNode)
 
+	flags.Duration(option.ARPPingRefreshPeriod, 5*time.Minute, "Period for ARP entries to remote nodes refresher (set 0 to disable)")
```
Slightly confusing description to me. Perhaps better phrased as something like "Remote node ARP entry refresh interval" or "Period for remote node ARP entry refresh"?

joamaki commented on pkg/datapath/linux/node.go:
```diff
@@ -666,46 +668,46 @@ func (n *linuxNodeHandler) getSrcAndNextHopIPv4(nodeIPv4 net.IP) (srcIPv4, nextH
-	if newNode.IsLocal() {
+	var link netlink.Link
+	n.neighLock.Lock()
+	if n.neighDiscoveryLink == nil || reflect.ValueOf(n.neighDiscoveryLink).IsNil() {
```
Out of curiosity.. Why is `ValueOf(...).IsNil()` needed?

brb replied:
https://mangatmodi.medium.com/go-check-nil-interface-the-right-way-d142776edef1

joamaki commented on pkg/datapath/linux/node.go:
```diff
@@ -719,25 +721,34 @@ func (n *linuxNodeHandler) insertNeighbor(ctx context.Context, newNode *nodeType
-	if !refresh {
-		nextHopIsNew = n.neighNextHopRefCount.Add(nextHopStr)
-	}
+	n.neighLock.Unlock() // to allow concurrent arpings bellow
```
```suggestion
	n.neighLock.Unlock() // to allow concurrent arpings below
```
